src/execution_plan/ops/op_semi_apply.c

```diff
@@ -62,19 +60,22 @@ Record SemiApplyConsume(OpBase *opBase) {
 		// Try to get a record from bound stream.
 		op->r = OpBase_Consume(op->bound_branch);
 		if(!op->r) return NULL; // Depleted.
+		// Propagate Record to the top of the Match stream.
+		if(op->op_arg) Argument_AddRecord(op->op_arg, OpBase_CloneRecord(op->r));
 
-		// Try to get a record from match stream.
-		Record righthand_record = _pullFromMatchStream(op);
-		if(righthand_record) {
+		bool matched = false;
+		Record rhs_record;
+		// Retrieve Records from the Match stream until it's depleted.
+		while((rhs_record = _pullFromMatchStream(op))) {
```

Comment:
I think it will be much more efficient to reset the stream once a single record is returned, thus skipping generating all other records.